### PR TITLE
insights: various critical bug fixes & improvements

### PR DESCRIPTION
**Pull request by slimsag**
1. Make the backend store insights per-repository. The historical enqueuer does this today, but not the regular enqueuer because we needed to extract out match counts per repository and this is a bit annoying to do. Now implemented. Helps #18966 and makes it clear how we'd implement querying of per-repo datapoints this in the future.

2. When considering backfilling data for certain timeframe, check if _the specific repo we're backfilling data for has any_. This should hopefully fix the issue on Sourcegraph.com and k8s.sgdev.org where the graph is totally flat which is very incorrect. Suspicion is that it managed to generate one data point correctly, ran into an error, restarted, and now thinks "we have data for those timeframes" Helps #18961

3. Ignore cached repos that are not found, e.g. if someone actually removes a repo and we have it in our cache to iterate over ignore it. Fixes an issue Thorsten noticed https://sourcegraph.slack.com/archives/CHPC7UX16/p1615468766294800

### CODE REVIEW

File: enterprise/internal/insights/background/queryrunner/work_handler.go [hunk 1]
```diff
@@ -84,27 +88,62 @@ func (r *workHandler) Handle(ctx context.Context, workerStore dbworkerstore.Stor
...
+	// results.
+	matchesPerRepo := map[string]int{}
+	for _, result := range results.Data.Search.Results.Results {
+		decoded, err := decodeResult(result)
+		if err != nil {
+			return errors.Wrap(err, fmt.Sprintf(`for query "%s"`, job.SearchQuery))
+		}
+		switch r := decoded.(type) {
+		case *fileMatch:
+			matchesPerRepo[r.Repository.ID] = matchesPerRepo[r.Repository.ID] + r.matchCount()
```
Comment (efritz) [inline enterprise/internal/insights/background/queryrunner/work_handler.go:112]: If you added a `key` method to the interface type you defined to bundle these types together, you could simplify this type switch into a generic method call to get the key to increment.

**Review by efritz — APPROVED** [review]

File: enterprise/internal/insights/background/queryrunner/work_handler.go [hunk 2]
```diff
@@ -84,27 +88,62 @@ func (r *workHandler) Handle(ctx context.Context, workerStore dbworkerstore.Stor
...
-		Point: store.SeriesPoint{
-			Time:  recordTime,
-			Value: float64(matchCount),
-		},
-		// TODO(slimsag): future: determine match count per repository, and store RepoID/RepoName (and maybe Metadata?)
-	})
+
+	// Figure out how many matches we got for every unique repository returned in the search
+	// results.
+	matchesPerRepo := map[string]int{}
```
[the remark this answers is not in the document]
Comment (slimsag) [reply enterprise/internal/insights/background/queryrunner/work_handler.go:104]:
```suggestion
	matchesPerRepo := make(map[string]int, len(results.Data.Search.Results.Results)*2)
```

I've added times 4 here to indicate that matches are likely > results.

File: enterprise/internal/insights/background/queryrunner/work_handler.go (hunk 1 above)
Comment (slimsag) [reply enterprise/internal/insights/background/queryrunner/work_handler.go:112]: Good call, done.

File: enterprise/internal/insights/background/queryrunner/graphql_results.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+import (
+	"encoding/json"
+	"fmt"
+)
+
+// TODO(slimsag): future: It's really nasty that our GraphQL search API is like this:
+//
+// 1. GraphQL makes it really annoying for us to extract the individual result values out
+//    due to the `... on Type` switches.
+// 2. More annoying, the GraphQL API has "results" which contain multiple matches. The web UI
```
[the remark this answers is not in the document]
Comment (slimsag) [reply enterprise/internal/insights/background/queryrunner/graphql_results.go:12]: I hope you're ready for some real gremlins and terror :) Here you go: https://github.com/sourcegraph/sourcegraph/pull/19082/commits/acd5bef7da70939cef2780613276629f22a0c2cb

File: enterprise/internal/insights/background/queryrunner/graphql_results.go [hunk 4]
```diff
@@ -15,6 +15,60 @@ import (
 //    API (we need per-repository match counts) we have to calculate this information ourselves
 //    in the same way the search backend does (see the matchCount() methods) in order to get the
 //    correct value. Yuck.
+//
+// Consider this GraphQL query snippet, which is *required* and *the only way* to count the total
+// number of matches (or as our UI calls them, "results"):
+//
+// 	results {
```
Comment (felixfbecker) [inline enterprise/internal/insights/background/queryrunner/graphql_results.go:22]: Is there a reason we can't just query `matchCount`? That's what the extension-based search insights do: https://github.com/sourcegraph/sourcegraph-search-insights/blob/e4fb6e3bc57f21a50333b30ae4f272d761144dd5/src/search-insights.ts#L129-L133

Comment (slimsag) [reply enterprise/internal/insights/background/queryrunner/graphql_results.go:22]: That is not per-repository, which is what we need.

Comment (felixfbecker) [reply enterprise/internal/insights/background/queryrunner/graphql_results.go:22]: Makes sense!

**Pull request merged**